pkg/signup/service/signup_service.go

```diff
@@ -116,9 +124,49 @@ func extractEmailHost(email string) string {
 	return email[i+1:]
 }
 
+// Signup reactivates the deactivated UserSignup resource or creates a new one with the specified username and userID
+// if doesn't exist yet.
+func (s *ServiceImpl) Signup(ctx *gin.Context) (*v1alpha1.UserSignup, error) {
+	userID := ctx.GetString(context.SubKey)
+	// Retrieve UserSignup resource from the host cluster
+	userSignup, err := s.CRTClient().V1Alpha1().UserSignups().Get(userID)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// New Signup
+			return s.createUserSignup(ctx)
+		}
+		return nil, err
+	}
+
+	// Check UserSignup status to determine whether user signup is deactivated
+	signupCondition, found := condition.FindConditionByType(userSignup.Status.Conditions, v1alpha1.UserSignupComplete)
+	if found && signupCondition.Status == apiv1.ConditionTrue && signupCondition.Reason == v1alpha1.UserSignupUserDeactivatedReason {
+		// Signup is deactivated. We need to reactivate it
+		return s.reactivateUserSignup(ctx)
+	}
+
+	username := ctx.GetString(context.UsernameKey)
+	return nil, errors2.Errorf("unable to create UserSignup [id: %s; username: %s] because there is already an active UserSignup with such ID", userID, username)
+}
+
+// reactivateUserSignup reactivates the deactivated UserSignup resource with the specified username and userID
+func (s *ServiceImpl) reactivateUserSignup(ctx *gin.Context) (*v1alpha1.UserSignup, error) {
+	userSignup, err := s.newUserSignup(ctx)
```

Comment:
If I understand it correctly - we need to create the new version of UserSignup here and cannot use the retrieved from this line [132](https://github.com/codeready-toolchain/registration-service/pull/144/files#diff-044c9e6475638d41c551f545f3899df6a42b37489faa3c9f7eb87edfa87b79b0R132) just to be sure that the reactivated UserSignup:
* is not banned
* uses the latest annotations & labels and other fields

Is that correct? I would propose adding some comment to this line to make it clear.

Reply:
During re-activation we basically want to override the existing deactivated resource so we have a new fresh one as it was never existed. The main reason is that we don't really know what else can be stored in the resource which can affect the "new" resource. So, I would avoid patching the existing resource and would just replace it (put vs patch) by the "new" one.
Stuff like checking bans, phone verification, etc is a free bonus here.

Reply:
Added a comment here: https://github.com/codeready-toolchain/registration-service/pull/144/commits/21738ee1e7b205e57a3a4004713486f8134d48ce